src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,474 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+            dspTreeID(lclOp));
+    weight.DumpTotalWeight();
+
+    // Remove or add a mask conversion.
+    LclMasksUpdateLclVisitor ev(this, lclOp, stmt, weight.simdBaseJitType, weight.simdSize);
+    GenTree*                 root = stmt->GetRootNode();
+    ev.WalkTree(&root, nullptr);
+
+    if (ev.found)
+    {
+        JITDUMP("Updated %s V%02d at [%06u] to mask (%s conversion)\n", isStore ? "store" : "var", lclOp->GetLclNum(),
+                dspTreeID(lclOp), ev.addedConversion ? "added" : "removed");
+
+#ifdef DEBUG
+        if (verbose)
+        {
+            gtDispTree(ev.modifiedOp);
+        }
+#endif
+    }
+}
+
+#endif // TARGET_ARM64
+
+//------------------------------------------------------------------------
+// optLclMasks: Allow locals to be of Mask type
+//
+// At the C# level, Masks share the same type as a Vector. It's possible for the same
+// variable to be used as a mask or vector. Any APIs that return a mask must first convert
+// the value to a vector before storing it to a variable. Any uses of a variable as a mask
+// must first convert from vector before using it. In many cases this creates unnecessary
+// conversions. For variables that live outside the scope of the current method then the
+// conversions are required to ensure correctness. However, for local variables where the
+// scope is local to the current method, then it is possible to keep the value as a mask,
+// by updating all definitions and uses.
+//
+// In the common case it is expected that uses of masks are consistent - once a variable is
+// created as a mask it will continue to be used and updated as a mask.
+//
+// In the uncommon case, a variable may be created in one type, used as another and/or
+// updated to a different type.
+//
+// For example (the conversion is implicit)
+//   vector<int> x = _ConvertMaskToVector_(CreateMask());
+//   x = Add(x, y);
+//
+// To account for this, the pass uses a weighting. For each variable, if it is a local store,
+// then count the existing cost of every covert to/from mask. Also count the cost for
+// switching the variable to store as mask (this may include adding additional conversions
+// as well as removing). For each counted instance, take into account the number of
+// instructions in the conversion and the weight of the block.
+//
+// Local variables that are defined then immediately used just once may not be saved to a
+// store. Here a convert to to vector will be used by a convert to mask. These instances will
+// be caught in the lowering phase.
+//
+// This weighting does not account for re-definition. A variable may first be created as a
+// mask used as such, then later in the method defined as a vector and used as such from
+// then on. This can be worked around at the user level by encouraging users not to reuse
+// variable names.
+//
+// Returns:
+//    Suitable phase status
+//
+PhaseStatus Compiler::fgOptimizeLclMasks()
```

Comment:
I don't think this should be in lclmorph.cpp. lclmorph.cpp is very specifically for the local morph phase that computes address exposure and rewrites `LCL_FLD` nodes. I would suggest a new file instead.